## 10.1 Panel Data

- In contrast to cross-section data where we have observations on n subjects (entities), panel data has observations on n entities at T ≥ 2 time periods.
- Where the index i refers to the entity while t refers to the time period, this is denoted

\begin{equation}
(X_{it},Y_{it}), \ i=1,\dots,n \ \ \ \text{and} \ \ \ t=1,\dots,T
\end{equation}

- Sometimes panel data is also called longitudinal data as it adds a temporal dimension to cross-sectional data. 
- Let us have a look at the dataset Fatalities by checking its structure and listing the first few observations.

In [3]:
using CSV #we use the CSV package to load the data
using DataFrames #we use the DataFrames package as the data is stored as an object of type "DataFrame"

fatalities = CSV.read("/mnt/juliabox/Econometrics With Julia/Datasets/fatalities.csv")

336×35 DataFrame. Omitted printing of 28 columns
│ Row │ Column1 │ state   │ year   │ spirits  │ unemp    │ income   │ emppop   │
│     │ Int64⍰  │ String⍰ │ Int64⍰ │ Float64⍰ │ Float64⍰ │ Float64⍰ │ Float64⍰ │
├─────┼─────────┼─────────┼────────┼──────────┼──────────┼──────────┼──────────┤
│ 1   │ 1       │ al      │ 1982   │ 1.37     │ 14.4     │ 10544.2  │ 50.692   │
│ 2   │ 2       │ al      │ 1983   │ 1.36     │ 13.7     │ 10732.8  │ 52.147   │
│ 3   │ 3       │ al      │ 1984   │ 1.32     │ 11.1     │ 11108.8  │ 54.1681  │
│ 4   │ 4       │ al      │ 1985   │ 1.28     │ 8.9      │ 11332.6  │ 55.2711  │
│ 5   │ 5       │ al      │ 1986   │ 1.23     │ 9.8      │ 11661.5  │ 56.5145  │
│ 6   │ 6       │ al      │ 1987   │ 1.18     │ 7.8      │ 11944.0  │ 57.5099  │
│ 7   │ 7       │ al      │ 1988   │ 1.17     │ 7.2      │ 12368.6  │ 56.8345  │
│ 8   │ 8       │ az      │ 1982   │ 1.97     │ 9.9      │ 12309.1  │ 56.8933  │
│ 9   │ 9       │ az      │ 1983   │ 1.9      │ 9.1      │ 12693.8  │ 57.5536  │
│ 10  │ 10      │ az      │ 1984   │ 2.14     │ 5.0      │ 13265.9  │ 60.379   │
⋮
│ 326 │ 326     │ wi      │ 1985   │ 2.02     │ 7.2      │ 13952.3  │ 61.6289  │
│ 327 │ 327     │ wi      │ 1986   │ 1.75     │ 7.0      │ 14351.9  │ 62.2111  │
│ 328 │ 328     │ wi      │ 1987   │ 1.68     │ 6.1      │ 14720.0  │ 64.7043  │
│ 329 │ 329     │ wi      │ 1988   │ 1.65     │ 4.3      │ 14941.3  │ 67.4431  │
│ 330 │ 330     │ wy      │ 1982   │ 2.19     │ 5.8      │ 14600.2  │ 67.3184  │
│ 331 │ 331     │ wy      │ 1983   │ 1.99     │ 8.4      │ 13574.5  │ 66.3912  │
│ 332 │ 332     │ wy      │ 1984   │ 1.91     │ 6.3      │ 13456.0  │ 66.0221  │
│ 333 │ 333     │ wy      │ 1985   │ 1.88     │ 7.1      │ 13595.3  │ 65.0838  │
│ 334 │ 334     │ wy      │ 1986   │ 1.7      │ 9.0      │ 13126.9  │ 63.9665  │
│ 335 │ 335     │ wy      │ 1987   │ 1.59     │ 8.6      │ 12719.0  │ 62.8571  │
│ 336 │ 336     │ wy      │ 1988   │ 1.55     │ 6.3      │ 13098.2  │ 64.6377  │

- We find that the dataset consists of 336 observations on 34 variables. 
- Notice that the variable state is a factor variable with 48 levels (one for each of the 48 contiguous federal states of the U.S.). 
- The variable year is also a factor variable that has 7 levels identifying the time period when the observation was made. 
- This gives us 7 × 48 = 336 observations in total. 
- Since all variables are observed for all entities and over all time periods, the panel is balanced. 
- If there were missing data for at least one entity in at least one time period we would call the panel unbalanced.

### Example: Traffic Deaths and Alcohol Taxes

- We start by reproducing Figure 10.1 of the book. 
- To this end we estimate simple regressions using data for years 1982 and 1988 that model the relationship between beer tax (adjusted for 1988 dollars) and the traffic fatality rate, measured as the number of fatalities per 10000 inhabitants. 
- Afterwards, we plot the data and add the corresponding estimated regression functions.
- If the kernel dies, try running the code using a local Julia environment.

In [ ]:
using FixedEffects #FixedEffectModels requires FixedEffects as a dependency
using FixedEffectModels #we use FixedEffectModels to create regression models
using CSV #we use the CSV package to load the data
using DataFrames #we use the DataFrames package as the data is stored as an object of type "DataFrame"
using Query #we use the Query package in order to filter the data using LINQ style query commands
using Plots 

fatalities = CSV.read("/mnt/juliabox/Econometrics With Julia/Datasets/fatalities.csv")

fatalities1982 = @from i in fatalities begin
    @where i.year == 1982
    @select {i.beertax, fatality_rate = i.fatal / i.pop * 10000}
    @collect DataFrame
end

fatalities1988 = @from i in fatalities begin
    @where i.year == 1988
    @select {i.beertax, fatality_rate = i.fatal / i.pop * 10000}
    @collect DataFrame
end

fatalities1982_model = reg( #initialise a FixedEffectModel and define it as fatalities1982_model
                        fatalities1982, #pass the DataFrame 'fatalities1982' as the dataset to be used in fatalities1982_model
                        @model(fatality_rate ~ beertax), #pass the regression formula consisting of the dependent variable 'fatality_rate' and the exogenous variable 'beertax'
)

fatalities1988_model = reg( #initialise a FixedEffectModel and define it as fatalities1988_model
                        fatalities1988, #pass the DataFrame 'fatalities1988' as the dataset to be used in fatalities1988_model
                        @model(fatality_rate ~ beertax), #pass the regression formula consisting of the dependent variable 'fatality_rate' and the exogenous variable 'beertax'
)

println(fatalities1982_model)
println(fatalities1988_model)



- The estimated regression functions are
\begin{align*}
  \widehat{FatalityRate} =& \, \underset{(0.15)}{2.01} + \underset{(0.13)}{0.15} \times BeerTax \quad (1982 \text{ data}), \\
  \widehat{FatalityRate} =& \, \underset{(0.11)}{1.86} + \underset{(0.13)}{0.44} \times BeerTax \quad (1988 \text{ data}).
\end{align*}

In [ ]:
x = fatalities1982.beertax
y = fatalities1982.fatality_rate

p1 = plot( #assign a plot object to the variable p1 using the following attributes
    x, #x series
    y, #y series
    st = :scatter, #series type
    title = "Traffic Fatality Rates and Beer Taxes in 1982", #plot title
    label = "US State", #legend labels
    xlabel = "Beer tax (in 1988 dollars)", #x axis label
    ylabel = "Fatality rate (fatalities per 10000)", #y axis label
    ylims = (0,4.5), #y axis limits
    yticks = 0:1:4.5, #y axis tick range
    ms = 4, #marker size
    mc = :blue #marker color
)

y_prediction_1982(xdata) = dot(coef(fatalities1982_model), [1, xdata]) #create a function that takes the dot product of the coefficients of the 'fatalities1982_model', 1 and the given x value 

x = [minimum(fatalities1982.beertax), maximum(fatalities1982.beertax)]
y = y_prediction_1982.(x) #broadcast the y function over the range of x values

plot!(
    p1,
    x,
    y,
    st = :line,
    label = "OLS"
)

display(p1)

x = fatalities1988.beertax
y = fatalities1988.fatality_rate

p2 = plot( #assign a plot object to the variable p2 using the following attributes
    x, #x series
    y, #y series
    st = :scatter, #series type
    title = "Traffic Fatality Rates and Beer Taxes in 1988", #plot title
    label = "US State", #legend labels
    xlabel = "Beer tax (in 1988 dollars)", #x axis label
    ylabel = "Fatality rate (fatalities per 10000)", #y axis label
    ylims = (0,4.5), #y axis limits
    yticks = 0:1:4.5, #y axis tick range
    ms = 4, #marker size
    mc = :blue #marker color
)

y_prediction_1988(xdata) = dot(coef(fatalities1988_model), [1, xdata]) #create a function that takes the dot product of the coefficients of the 'fatalities1988_model', 1 and the given x value

x = [minimum(fatalities1988.beertax), maximum(fatalities1988.beertax)] 
y = y_prediction_1988.(x)  #broadcast the y function over the range of x values

plot!(
    p2,
    x,
    y,
    st = :line,
    label = "OLS"
)

display(p2)
